tree/30_generic_methods/_package_state.cf

```diff
@@ -0,0 +1,260 @@
+# @documentation See package_present and package_absent for documentation.
+# @parameter name Name of the package, or path to a local package if state is present
+# @parameter version Version of the package, can be "latest" for latest version or "any" for any version (defaults to "any")
+# @parameter architecture Architecture of the package, can be an architecture name  or "default" (defaults to "default")
+# @parameter provider Package provider to use, can be "yum", "apt", "zypper", "pkg" or "default" for system default package manager (defaults to "default")
+# 
+# @class_prefix package_${state}
+# @class_parameter name
+
+bundle agent _package_state(name, version, architecture, provider, state)
+{
+
+  vars:
+      "canonified_name"  string => canonify("${name}");
+      "old_class_prefix" string => "package_${state}_${canonified_name}";
+      "promisers"        slist  => { @{this.callers_promisers}, cf_null }, policy => "ifdefined";
+      "class_prefix"     string => canonify(join("_", "promisers"));
+      "args"             slist  => { "${name}", "${version}", "${architecture}", "${provider}", "${state}" };
+
+    # Build string vars used for reporting
+
+    # State
+    state_present::
+      "state_description" string => "Presence";
+    !state_present::
+      "state_description" string => "Absence";
+
+    # Architecture
+    architecture_specified::
+      "architecture_description" string => " for ${architecture} architecture ";
+    !architecture_specified::
+      "architecture_description" string => " ";
+
+    # Version
+    version_latest::
+      "version_description" string => "in latest available version";
+    !version_specified::
+      "version_description" string => "in any version";
+    version_specified.!version_latest::
+      "version_description" string => "in version ${version}";
+
+  defaults:
+      "version"      string => "any", if_match_regex => "";
+      "architecture" string => "default", if_match_regex => "";
+      "provider"     string => "default", if_match_regex => "";
+      # Select the default packager for this platform if "default"
+      "provider"     string => "${package_module_knowledge.platform_default}", if_match_regex => "default";
+
+  classes:
+      "version_latest"  expression => strcmp("latest", "${version}");
+      # As "latest" is understood by package promises,
+      # the only special case is "any", which maps to
+      # no specified version in the promise
+      "version_specified"      not => strcmp("any", "${version}");
+
+      "state_present" expression => strcmp("present", "${state}");
+
+      # If architecture is not specified, do not add it to the promise.
+      # The package module will pick the default one for the local platform
+      "architecture_specified" not => strcmp("default", "${architecture}");
+
+      # Select the right package manager
+      # Always use the specified one, and select a default based on the OS
```

Comment:
because of the `"provider"     string => "${package_module_knowledge.platform_default}", if_match_regex => "default";` above.